lib/spack/spack/paths.py

```diff
@@ -9,13 +9,108 @@
 dependencies.
 """
 import os
+import pathlib
 from pathlib import PurePath
+from typing import Optional
 
 import llnl.util.filesystem
 
+import spack.util.hash as hash
+
 #: This file lives in $prefix/lib/spack/spack/__file__
 prefix = str(PurePath(llnl.util.filesystem.ancestor(__file__, 4)))
 
+xdg_config_home = "XDG_CONFIG_HOME"
+xdg_state_home = "XDG_STATE_HOME"
+xdg_data_home = "XDG_DATA_HOME"
+spack_data_home_varname = "SPACK_DATA_HOME"
+xdg_cache_home = "XDG_CACHE_HOME"
+
+
+# This is for tests that want to clean the environment of XDG_ variables that
+# affect spack behavior
+def _unset_xdg_vars(env):
+    saved = {}
+    for xdg_var in [xdg_config_home, xdg_state_home, xdg_data_home, xdg_cache_home]:
+        if xdg_var in env:
+            saved[xdg_var] = env.pop(xdg_var)
+    return saved
+
+
+def _define_xdg_or_backup(xdg_var, backup):
+    if xdg_var in os.environ:
+        spack_xdg_defined = os.path.join(os.environ[xdg_var], "spack")
+    else:
+        spack_xdg_defined = os.path.join(backup, "spack")
+    return os.path.expanduser(spack_xdg_defined)
```

Comment:
```suggestion
def _define_xdg_or_backup(xdg_base, backup):
    # xdg_base = {DATA_HOME, CONFIG_HOME, STATE_HOME, etc}
    spack_override = f"SPACK_{xdg_base}"
    xdg_var = f"XDG_{xdg_base}"
    if spack_override in os.environ:
        spack_xdg_defined = os.path.join(os.environ[spack_override], "spack")
    elif xdg_var in os.environ:
        spack_xdg_defined = os.path.join(os.environ[xdg_var], "spack")
    else:
        spack_xdg_defined = os.path.join(backup, "spack")
    return os.path.expanduser(spack_xdg_defined)
```

Can we just support an override of all XDG variables? I think that simplifies the precedent logic to the users. `config > SPACK_* > XDG_*`

Then we can delete the `spack_data_home_varname` variable and special logic for respecting it.

Reply:
I wanted to keep track of e.g. XDG_DATA_HOME and SPACK_DATA_HOME separately for https://github.com/spack/spack/pull/47615#discussion_r2125757909, but if they end up "adjacent" in that list, then this implementation is good.